README.md
<div align="center">

# CC-Leaderboards

A component that simplifies the process of adding a global leaderboard to a game.

[![Build Status](https://github.com/Core-Team-META/CC-Leaderboards/workflows/CI/badge.svg)](https://github.com/Core-Team-META/CC-Leaderboards/actions/workflows/ci.yml?query=workflow%3ACI%29)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/Core-Team-META/CC-Leaderboards?style=plastic)

![TitleCard](/Screenshots/TitleCard.png)

</div>

## Overview

Leaderboards is a component that simplifies the process of adding a global leaderboard to a game. It is as simple as dragging and dropping a leaderboard template into the hierarchy and adjusting to meet your needs.

Both World and UI versions supported.

Created by @NicholasForeman of Team META

Make sure to read the Leaderboards_README file for setup and configuration instructions

## Credits

Thanks to:

- @standardcombo for review and documentation template
- @Aggripina for UI thumbnail design


CHANGELOG.md
# Changelog

## v1.0.2

- Fix minor bug with UpdateOnRoundEnd

## v1.0.1

- Improved Documentation
- Fixed a bug with damage event

## v1.0.0

- Improved Readme


CommunityContent/META Leaderboards/Scripts/Leaderboard.lua
--[[

	Leaderboards - Server
	1.0.1 - 2020/10/15
	by Nicholas Foreman (META) (https://www.coregames.com/user/f9df3457225741c89209f6d484d0eba8)

--]]

-- Make sure Leaderboard_DataTracker is in the hierarchy and we have access to the ResourcesToTrack global table
do local startTime = time()
	while(not (type(_G.ResourcesToTrack) == "table")) do
		Task.Wait()

		if((time() - startTime) > 3) then
			error("An instance of Leaderboard_DataTracker is not in the hierarchy")
		end
	end
end

------------------------------------------------------------------------------------------------------------------------
--	OBJECTS AND REFERENCES
------------------------------------------------------------------------------------------------------------------------
local Leaderboard = script:GetCustomProperty("Leaderboard"):WaitForObject()

local LeaderboardReference = Leaderboard:GetCustomProperty("LeaderboardReference")
assert(LeaderboardReference.isAssigned, "The NetReference provided is not properly set for LeaderboardReference.")

------------------------------------------------------------------------------------------------------------------------
--	CONSTANTS
------------------------------------------------------------------------------------------------------------------------
local LEADERBOARD_TYPE = Leaderboard:GetCustomProperty("LeaderboardType")
local LEADERBOARD_STAT = Leaderboard:GetCustomProperty("LeaderboardStat")
local LEADERBOARD_PERSISTENCE = Leaderboard:GetCustomProperty("LeaderboardPersistence")

-- Only applicable if LEADERBOARD_STAT is RESOURCE
local RESOURCE_NAME = Leaderboard:GetCustomProperty("ResourceName")

local UPDATE_ON_RESOURCE_CHANGED = Leaderboard:GetCustomProperty("UpdateOnResourceChanged")
local UPDATE_ON_PLAYER_DIED = Leaderboard:GetCustomProperty("UpdateOnPlayerDied")
local UPDATE_ON_DAMAGE_DEALT = Leaderboard:GetCustomProperty("UpdateOnDamageDealt")
local UPDATE_ON_EVENT = Leaderboard:GetCustomProperty("UpdateOnEvent")

local LEADERBOARD_TYPES = { "GLOBAL", "MONTHLY", "WEEKLY", "DAILY" }
local LEADERBOARD_STATS = { "RESOURCE", "KDR", "KILLS", "DEATHS", "DAMAGE_DEALT" }
local LEADERBOARD_PERSISTENCES = { "TOTAL", "ROUND" }

------------------------------------------------------------------------------------------------------------------------
--	LOCAL FUNCTIONS
------------------------------------------------------------------------------------------------------------------------

local function OnUpdate(leaderboardPersistence, leaderboardStat, player, score, resourceName)
	if(leaderboardPersistence ~= LEADERBOARD_PERSISTENCE) then return end

	if(leaderboardStat == "KDR") then
		if(LEADERBOARD_STAT ~= "KDR") then return end

		Leaderboards.SubmitPlayerScore(LeaderboardReference, player, score)

		if(UPDATE_ON_PLAYER_DIED) then
			Update()
		end
	elseif(leaderboardStat == "KILLS") then
		if(LEADERBOARD_STAT ~= "KILLS") then return end

		Leaderboards.SubmitPlayerScore(LeaderboardReference, player, score)

		if(UPDATE_ON_PLAYER_DIED) then
			Update()
		end
	elseif(leaderboardStat == "DEATHS") then
		if(LEADERBOARD_STAT ~= "DEATHS") then return end

		Leaderboards.SubmitPlayerScore(LeaderboardReference, player, score)

		if(UPDATE_ON_PLAYER_DIED) then
			Update()
		end
	elseif(leaderboardStat == "DAMAGE_DEALT") then
		if(LEADERBOARD_STAT ~= "DAMAGE_DEALT") then return end

		Leaderboards.SubmitPlayerScore(LeaderboardReference, player, score)

		if(UPDATE_ON_DAMAGE_DEALT) then
			Update()
		end
	else
		if(LEADERBOARD_STAT ~= "RESOURCE") then return end
		if(#RESOURCE_NAME <= 0) then return end
		if(resourceName ~= RESOURCE_NAME) then return end

		Leaderboards.SubmitPlayerScore(LeaderboardReference, player, score)

		if(UPDATE_ON_RESOURCE_CHANGED) then
			Update()
		end
	end
end

local function GetProperty(value, options)
	value = string.upper(value)

	for _, option in pairs(options) do
		if(value == option) then return value end
	end

	return options[1]
end

function Update()
	while(Events.BroadcastToAllPlayers("LDT_Update", Leaderboard.id) == BroadcastEventResultCode.EXCEEDED_RATE_LIMIT) do
		Task.Wait()
	end
end

LEADERBOARD_TYPE = GetProperty(LEADERBOARD_TYPE, LEADERBOARD_TYPES)
LEADERBOARD_STAT = GetProperty(LEADERBOARD_STAT, LEADERBOARD_STATS)
LEADERBOARD_PERSISTENCE = GetProperty(LEADERBOARD_PERSISTENCE, LEADERBOARD_PERSISTENCES)

if((LEADERBOARD_STAT == "RESOURCE") and (#RESOURCE_NAME > 0)) then
	_G.ResourcesToTrack[RESOURCE_NAME] = true
end

Events.Connect("LDT_Update", OnUpdate)

if(#UPDATE_ON_EVENT > 0) then
	Events.Connect(UPDATE_ON_EVENT, Update)
end

Data/Subfolders/META Leaderboards/Scripts/Leaderboard_World.lua
--[[

	Leaderboards - World (Client)
	1.0.1 - 2020/10/15
	by Nicholas Foreman (META) (https://www.coregames.com/user/f9df3457225741c89209f6d484d0eba8)

--]]

------------------------------------------------------------------------------------------------------------------------
--	OBJECTS AND REFERENCES
------------------------------------------------------------------------------------------------------------------------
local EntryTemplate = script:GetCustomProperty("EntryTemplate")
local Leaderboard = script:GetCustomProperty("Leaderboard"):WaitForObject()

local LeaderboardReference = Leaderboard:GetCustomProperty("LeaderboardReference")
assert(LeaderboardReference.isAssigned, "The NetReference provided is not properly set for LeaderboardReference.")

local Entries = script:GetCustomProperty("Entries"):WaitForObject()
local Title = script:GetCustomProperty("Title"):WaitForObject()
local UpdateTimer = script:GetCustomProperty("UpdateTimer"):WaitForObject()

------------------------------------------------------------------------------------------------------------------------
--	CONSTANTS
------------------------------------------------------------------------------------------------------------------------
local LEADERBOARD_TYPE = Leaderboard:GetCustomProperty("LeaderboardType")
local LEADERBOARD_STAT = Leaderboard:GetCustomProperty("LeaderboardStat")
local LEADERBOARD_PERSISTENCE = Leaderboard:GetCustomProperty("LeaderboardPersistence")

-- Only applicable if LEADERBOARD_STAT is RESOURCE
local RESOURCE_NAME = Leaderboard:GetCustomProperty("ResourceName")

local DISPLAY_AS_INTEGER = Leaderboard:GetCustomProperty("DisplayAsInteger")

local UPDATE_ON_ROUND_END = Leaderboard:GetCustomProperty("UpdateOnRoundEnd")
local UPDATE_ON_EVENT = Leaderboard:GetCustomProperty("UpdateOnEvent")
local UPDATE_TIMER = math.abs(Leaderboard:GetCustomProperty("UpdateTimer"))

local FIRST_PLACE_COLOR = Leaderboard:GetCustomProperty("FirstPlaceColor")
local SECOND_PLACE_COLOR = Leaderboard:GetCustomProperty("SecondPlaceColor")
local THIRD_PLACE_COLOR = Leaderboard:GetCustomProperty("ThirdPlaceColor")
local NO_PODIUM_PLACEMENT_COLOR = Leaderboard:GetCustomProperty("NoPodiumPlacementColor")
local USERNAME_COLOR = Leaderboard:GetCustomProperty("UsernameColor")
local SCORE_COLOR = Leaderboard:GetCustomProperty("ScoreColor")

local LEADERBOARD_TYPES = { "GLOBAL", "MONTHLY", "WEEKLY", "DAILY" }
local LEADERBOARD_STATS = { "RESOURCE", "KDR", "KILLS", "DEATHS", "DAMAGE_DEALT" }
local LEADERBOARD_PERSISTENCES = { "TOTAL", "ROUND" }

local SCORE_SUFFIXES = {"K", "M", "B", "T", "Q", "Qu", "S", "Se", "O", "N", "D"}

------------------------------------------------------------------------------------------------------------------------
--	INITIAL VARIABLES
------------------------------------------------------------------------------------------------------------------------
local currentEntries = {}
local lastUpdate = time()

------------------------------------------------------------------------------------------------------------------------
--	LOCAL FUNCTIONS
------------------------------------------------------------------------------------------------------------------------

local function ToSuffixString(num)
	for index = #SCORE_SUFFIXES, 1, -1 do
		local value = 10 ^ (index * 3)
		if num >= value then
			return string.format("%.1f", num / value) .. SCORE_SUFFIXES[index]
		end
	end
	return string.format("%.1f", num)
end

local function GetShouldUpdate(newEntries, currentEntries)
	if(#newEntries ~= #currentEntries) then return true end

	for index, newEntry in ipairs(newEntries) do
		if(index > 10) then break end

		local currentEntry = currentEntries[index]
		if(not currentEntry) then return true end

		if(newEntry.id ~= currentEntry.id) then
			return true
		elseif(newEntry.score ~= currentEntry.score) then
			return true
		end
	end

	return false
end

local function GetFirstTenEntries(entries)
	local newEntries = {}

	for index, entry in ipairs(entries) do
		if(index > 10) then break end

		newEntries[index] = entries[index]
	end

	return newEntries
end

local function ClearEntries()
	for _, child in pairs(Entries:GetChildren()) do
		child:Destroy()
	end
end

local function GetProperty(value, options)
	value = string.upper(value)

	for _, option in pairs(options) do
		if(value == option) then return value end
	end

	return options[1]
end

------------------------------------------------------------------------------------------------------------------------
--	GLOBAL FUNCTIONS
------------------------------------------------------------------------------------------------------------------------

function Update(id)
	if(not Leaderboards.HasLeaderboards()) then return end

	if(id) then
		if(id ~= Leaderboard.id) then return end
	end

	local newEntries = Leaderboards.GetLeaderboard(LeaderboardReference, LeaderboardType[LEADERBOARD_TYPE])
	if(not newEntries) then Task.Wait() return Update() end
	newEntries = GetFirstTenEntries(newEntries)

	local shouldUpdate = GetShouldUpdate(newEntries, currentEntries)
	if(not shouldUpdate) then return end

	ClearEntries()
	currentEntries = newEntries

	for index, entry in ipairs(newEntries) do
		local leaderboardEntry = World.SpawnAsset(EntryTemplate, {
			parent = Entries,
		})
		leaderboardEntry.name = entry.name

		local playerPosition, playerName, playerScore =
			leaderboardEntry:GetCustomProperty("Position"):WaitForObject(),
			leaderboardEntry:GetCustomProperty("Name"):WaitForObject(),
			leaderboardEntry:GetCustomProperty("Score"):WaitForObject()

		playerPosition.text = string.format("%002i", index)
		if(index == 1) then
			playerPosition:SetColor(FIRST_PLACE_COLOR)
		elseif(index == 2) then
			playerPosition:SetColor(SECOND_PLACE_COLOR)
		elseif(index == 3) then
			playerPosition:SetColor(THIRD_PLACE_COLOR)
		else
			playerPosition:SetColor(NO_PODIUM_PLACEMENT_COLOR)
		end

		playerName.text = string.sub(entry.name, 1, 23)
		playerName:SetColor(USERNAME_COLOR)

		if(DISPLAY_AS_INTEGER) then
			playerScore.text = tostring(math.ceil(entry.score))
		else
			playerScore.text = ToSuffixString(entry.score)
		end
		playerScore:SetColor(SCORE_COLOR)

		leaderboardEntry:SetPosition(Vector3.New(0, 0, -30 * (index - 1)))
	end
end

function Tick()
	if((time() - lastUpdate) < UPDATE_TIMER) then
		local timeLeft = math.ceil((lastUpdate + UPDATE_TIMER) - time())
		UpdateTimer.text = string.format("UPDATES IN %s SECONDS", timeLeft)
	else
		lastUpdate = time()

		Update()
	end
end

------------------------------------------------------------------------------------------------------------------------
--	INITIALIZATION
------------------------------------------------------------------------------------------------------------------------

Events.Connect("LDT_Update", Update)

if(UPDATE_ON_ROUND_END) then
	Game.roundEndEvent:Connect(Update)
end

if(#UPDATE_ON_EVENT > 0) then
	Events.Connect(UPDATE_ON_EVENT, Update)
end

LEADERBOARD_TYPE = GetProperty(LEADERBOARD_TYPE, LEADERBOARD_TYPES)
LEADERBOARD_STAT = GetProperty(LEADERBOARD_STAT, LEADERBOARD_STATS)
LEADERBOARD_PERSISTENCE = GetProperty(LEADERBOARD_PERSISTENCE, LEADERBOARD_PERSISTENCES)

if(UPDATE_TIMER <= 0) then
	UPDATE_TIMER = 0
end

if(UPDATE_TIMER <= 0) then
	UPDATE_TIMER = 0
end

if(LEADERBOARD_STAT == "RESOURCE") then
	Title.text = string.format("%s %s %s", LEADERBOARD_TYPE, LEADERBOARD_PERSISTENCE, string.upper(RESOURCE_NAME))
else
	Title.text = string.format("%s %s %s", LEADERBOARD_TYPE, LEADERBOARD_PERSISTENCE, LEADERBOARD_STAT)
end

while(not Leaderboards.HasLeaderboards()) do
	Task.Wait()
end
Task.Wait()
Update()

Data/Subfolders/META Leaderboards/Scripts/Leaderboards_README.lua
--[[

	Leaderboards - README
	1.0.2 - 2021/02/01
	by Nicholas Foreman (META) (https://www.coregames.com/user/f9df3457225741c89209f6d484d0eba8)

	Leaderboards is a component that simplifies the process of adding a global leaderboard to a game. It is as simple as dragging
	and dropping a leaderboard template into the hierarchy and adjusting to meet your needs.


1.	Creating a Global Leaderboard

	1)	Go to the "Global Leaderboards" tab in the editor: View > Global Leaderboards
	2)	Click "Create New Leaderboard"
	3)	Follow the prompt to create your leaderboard


1b.	Setup

	1)	Drag and drop "Leaderboard Dependencies" into the hierarchy
		Note: Make sure this is the only version in the hierarchy
	2)	Drag and drop either a "World Leaderboard" or "Interface Leaderboard" into the hierarchy
	3)	With the Global Leaderboards view still open, select the World/Interface Leaderboard object and locate the LeaderboardReference
		custom property. It should say *Missing Reference*. Drag your leaderboard from the Global Leaderboards view onto the Properties
		view to assign the custom property.
	4)	Alter custom properties for each template as needed.


2.	Usage

	The most basic usage includes KDR, Kills, and Deaths by dragging the template (world or interface) into the hierarchy and
	adjusting LeaderboardStat to be KDR, KILLS, or DEATHS. However, more advanced usage includes Resources in which you can
	set LeaderboardStat to be RESOURCE and change ResourceName to any resource of your choosing, such as "Money".


3.	Leaderboard Dependencies

	"Leaderboard Dependencies" is a single template that must be in the hierarchy and remain in the hierarchy. There should
	only be one of this template in the hierarchy. It handles the persistent tracking of resources, kills, deaths, and kdr.
	Having multiple of this in the hierarchy can cause unwanted bugs and possible corruption. Not including it in the hierarchy
	will make each of the leaderboard template not function, either.


3b.	World Leaderboard

	A "World Leaderboard" is a physical panel in the world utilizing WorldText to display rankings.


3c.	Interface Leaderboard

	An "Interface Leaderboard" is a UI panel utilizing UI elements to display rankings on the player's screen. With slightly
	more customization, included is the ability to toggle and ease in/out the UI.
	

4. Monthly, Weekly or Daily Time Frames

	1) Leaderboards that show Monthly, Weekly and Daily scores can be created with a single entry in the Global Leaderboards menu.
	2) For each desired time frame, add a copy of the World/Interface templates, and assign the same leaderboard entry to each one.
	3) Then, for each template, configure the "LeaderboardType" custom property to say GLOBAL, MONTHLY, WEEKLY or DAILY.


5.	Discord

	If you have any questions, feel free to join the Core Hub Discord server:
		discord.gg/core-creators
	We are a friendly group of creators and players in the Core community. Everyone is welcome to play games together or
	learn about game dev!

--]]



Data/Subfolders/META Leaderboards/Scripts/Leaderboard_Interface.lua
--[[

	Leaderboards - Interface (Client)
	1.0.1 - 2020/10/15
	by Nicholas Foreman (META) (https://www.coregames.com/user/f9df3457225741c89209f6d484d0eba8)

--]]

------------------------------------------------------------------------------------------------------------------------
--	EXTERNAL SCRIPTS AND APIS
------------------------------------------------------------------------------------------------------------------------
local EaseUI = require(script:GetCustomProperty("EaseUI"))

------------------------------------------------------------------------------------------------------------------------
--	OBJECTS AND REFERENCES
------------------------------------------------------------------------------------------------------------------------
local LocalPlayer = Game.GetLocalPlayer()

local EntryTemplate = script:GetCustomProperty("EntryTemplate")
local Leaderboard = script:GetCustomProperty("Leaderboard"):WaitForObject()

local LeaderboardReference = Leaderboard:GetCustomProperty("LeaderboardReference")
assert(LeaderboardReference.isAssigned, "The NetReference provided is not properly set for LeaderboardReference.")

local LeaderboardPanel = script:GetCustomProperty("LeaderboardPanel"):WaitForObject()
local Entries = script:GetCustomProperty("Entries"):WaitForObject()
local Title = script:GetCustomProperty("Title"):WaitForObject()
local UpdateTimer = script:GetCustomProperty("UpdateTimer"):WaitForObject()

------------------------------------------------------------------------------------------------------------------------
--	CONTANTS
------------------------------------------------------------------------------------------------------------------------
local LEADERBOARD_TYPE = Leaderboard:GetCustomProperty("LeaderboardType")
local LEADERBOARD_STAT = Leaderboard:GetCustomProperty("LeaderboardStat")
local LEADERBOARD_PERSISTENCE = Leaderboard:GetCustomProperty("LeaderboardPersistence")

-- Only applicable if LEADERBOARD_STAT is RESOURCE
local RESOURCE_NAME = Leaderboard:GetCustomProperty("ResourceName")

local DISPLAY_AS_INTEGER = Leaderboard:GetCustomProperty("DisplayAsInteger")

local UPDATE_ON_ROUND_END = Leaderboard:GetCustomProperty("UpdateOnRoundEnd")
local UPDATE_ON_EVENT = Leaderboard:GetCustomProperty("UpdateOnEvent")
local UPDATE_TIMER = math.abs(Leaderboard:GetCustomProperty("UpdateTimer"))

local FIRST_PLACE_COLOR = Leaderboard:GetCustomProperty("FirstPlaceColor")
local SECOND_PLACE_COLOR = Leaderboard:GetCustomProperty("SecondPlaceColor")
local THIRD_PLACE_COLOR = Leaderboard:GetCustomProperty("ThirdPlaceColor")
local NO_PODIUM_PLACEMENT_COLOR = Leaderboard:GetCustomProperty("NoPodiumPlacementColor")
local USERNAME_COLOR = Leaderboard:GetCustomProperty("UsernameColor")
local SCORE_COLOR = Leaderboard:GetCustomProperty("ScoreColor")

local TOGGLE_BINDING = Leaderboard:GetCustomProperty("ToggleBinding")
local TOGGLE_EVENT = Leaderboard:GetCustomProperty("ToggleEvent")
local FORCE_ON_EVENT = Leaderboard:GetCustomProperty("ForceOnEvent")
local FORCE_OFF_EVENT = Leaderboard:GetCustomProperty("ForceOffEvent")

local EASE_TOGGLE = Leaderboard:GetCustomProperty("EaseToggle")
local EASE_BEGINNING = Leaderboard:GetCustomProperty("EaseBeginning")
local EASING_DURATION = Leaderboard:GetCustomProperty("EasingDuration")
local EASING_EQUATION_IN = Leaderboard:GetCustomProperty("EasingEquationIn")
local EASING_DIRECTION_IN = Leaderboard:GetCustomProperty("EasingDirectionIn")
local EASING_EQUATION_OUT = Leaderboard:GetCustomProperty("EasingEquationOut")
local EASING_DIRECTION_OUT = Leaderboard:GetCustomProperty("EasingDirectionOut")

local LEADERBOARD_TYPES = { "GLOBAL", "MONTHLY", "WEEKLY", "DAILY" }
local LEADERBOARD_STATS = { "RESOURCE", "KDR", "KILLS", "DEATHS", "DAMAGE_DEALT" }
local LEADERBOARD_PERSISTENCES = { "TOTAL", "ROUND" }
local EASE_BEGINNINGS = { "UP", "DOWN", "LEFT", "RIGHT" }

local SCORE_SUFFIXES = { "K", "M", "B", "T", "Q", "Qu", "S", "Se", "O", "N", "D" }

------------------------------------------------------------------------------------------------------------------------
--	INITIAL VARIABLES
------------------------------------------------------------------------------------------------------------------------
local currentEntries = {}
local lastUpdate = time()
local isVisible = false

local lastTask

local isCursorVisible, canCursorInteractWithUI =
	UI.IsCursorVisible(), UI.CanCursorInteractWithUI()

------------------------------------------------------------------------------------------------------------------------
--	LOCAL FUNCTIONS
------------------------------------------------------------------------------------------------------------------------

local function ToSuffixString(num)
	for index = #SCORE_SUFFIXES, 1, -1 do
		local value = 10 ^ (index * 3)
		if num >= value then
			return string.format("%.1f", num / value) .. SCORE_SUFFIXES[index]
		end
	end
	return string.format("%.1f", num)
end

local function GetShouldUpdate(newEntries, currentEntries)
	if(#newEntries ~= #currentEntries) then return true end

	for index, newEntry in ipairs(newEntries) do
		if(index > 10) then break end

		local currentEntry = currentEntries[index]
		if(not currentEntry) then return true end

		if(newEntry.id ~= currentEntry.id) then
			return true
		elseif(newEntry.score ~= currentEntry.score) then
			return true
		end
	end

	return false
end

local function GetFirstTenEntries(entries)
	local newEntries = {}

	for index, entry in ipairs(entries) do
		if(index > 10) then break end

		newEntries[index] = entries[index]
	end

	return newEntries
end

local function ClearEntries()
	for _, child in pairs(Entries:GetChildren()) do
		child:Destroy()
	end
end

local function GetProperty(value, options)
	value = string.upper(value)

	for _, option in pairs(options) do
		if(value == option) then return value end
	end

	return options[1]
end

local function OnBindingReleased(player, binding)
	if(binding ~= TOGGLE_BINDING) then return end

	ForceToggle()
end

------------------------------------------------------------------------------------------------------------------------
--	GLOBAL FUNCTIONS
------------------------------------------------------------------------------------------------------------------------

function ForceOn()
	isVisible = true

	isCursorVisible, canCursorInteractWithUI =
		UI.IsCursorVisible(), UI.CanCursorInteractWithUI()

	UI.SetCursorVisible(true)
	UI.SetCanCursorInteractWithUI(true)

	LeaderboardPanel.visibility = Visibility.FORCE_ON
	if(EASE_TOGGLE) then
		local vertical = (EASE_BEGINNING == "UP") or (EASE_BEGINNING == "DOWN")
		if(vertical) then
			EaseUI.EaseY(LeaderboardPanel, 0, EASING_DURATION, EASING_EQUATION_IN, EASING_DIRECTION_IN)
		else
			EaseUI.EaseX(LeaderboardPanel, 0, EASING_DURATION, EASING_EQUATION_IN, EASING_DIRECTION_IN)
		end
	end
end

function ForceOff()
	isVisible = false

	UI.SetCursorVisible(isCursorVisible)
	UI.SetCanCursorInteractWithUI(canCursorInteractWithUI)

	isCursorVisible = nil
	canCursorInteractWithUI = nil

	if(EASE_TOGGLE) then
		if(EASE_BEGINNING == "UP") then
			EaseUI.EaseY(LeaderboardPanel, -1500, EASING_DURATION, EASING_EQUATION_OUT, EASING_DIRECTION_OUT)
		elseif(EASE_BEGINNING == "DOWN") then
			EaseUI.EaseY(LeaderboardPanel, 1500, EASING_DURATION, EASING_EQUATION_OUT, EASING_DIRECTION_OUT)
		elseif(EASE_BEGINNING == "LEFT") then
			EaseUI.EaseX(LeaderboardPanel, -1500, EASING_DURATION, EASING_EQUATION_OUT, EASING_DIRECTION_OUT)
		elseif(EASE_BEGINNING == "RIGHT") then
			EaseUI.EaseX(LeaderboardPanel, 1500, EASING_DURATION, EASING_EQUATION_OUT, EASING_DIRECTION_OUT)
		end

		local task
		task = Task.Spawn(function()
			Task.Wait(EASING_DURATION)

			if((not lastTask) or (lastTask ~= task)) then return end
			lastTask = nil

			if(not isVisible) then
				LeaderboardPanel.visibility = Visibility.FORCE_OFF
			end
		end)
		lastTask = task
	else
		LeaderboardPanel.visibility = Visibility.FORCE_OFF
	end
end

function ForceToggle()
	if(isVisible) then
		ForceOff()
	else
		ForceOn()
	end
end

function Toggle(id)
	if(id) then
		if(id ~= Leaderboard.id) then return end
	end

	ForceToggle()
end

function Update(id)
	if(not Leaderboards.HasLeaderboards()) then return end

	if(id) then
		if(id ~= Leaderboard.id) then return end
	end

	local newEntries = Leaderboards.GetLeaderboard(LeaderboardReference, LeaderboardType[LEADERBOARD_TYPE])
	if(not newEntries) then return end
	newEntries = GetFirstTenEntries(newEntries)

	local shouldUpdate = GetShouldUpdate(newEntries, currentEntries)
	if(not shouldUpdate) then return end

	ClearEntries()
	currentEntries = newEntries

	local count = #newEntries
	for index, entry in ipairs(newEntries) do
		local leaderboardEntry = World.SpawnAsset(EntryTemplate, {
			parent = Entries,
		})
		leaderboardEntry.name = entry.name

		local playerPosition, playerName, playerScore =
			leaderboardEntry:GetCustomProperty("Position"):WaitForObject(),
			leaderboardEntry:GetCustomProperty("Name"):WaitForObject(),
			leaderboardEntry:GetCustomProperty("Score"):WaitForObject()

		playerPosition.text = string.format("%002i", index)
		if(index == 1) then
			playerPosition:SetColor(FIRST_PLACE_COLOR)
		elseif(index == 2) then
			playerPosition:SetColor(SECOND_PLACE_COLOR)
		elseif(index == 3) then
			playerPosition:SetColor(THIRD_PLACE_COLOR)
		else
			playerPosition:SetColor(NO_PODIUM_PLACEMENT_COLOR)
		end

		playerName.text = string.sub(entry.name, 1, 30)
		playerName:SetColor(USERNAME_COLOR)

		if(DISPLAY_AS_INTEGER) then
			playerScore.text = tostring(math.ceil(entry.score))
		else
			playerScore.text = ToSuffixString(entry.score)
		end
		playerScore:SetColor(SCORE_COLOR)

		leaderboardEntry.y = (leaderboardEntry.height * (index - 1)) + (5 * (index - 1))

		if(count >= 8) then
			leaderboardEntry.width = leaderboardEntry.width - 20
		end
	end
end

function Tick()
	if((time() - lastUpdate) < UPDATE_TIMER) then
		local timeLeft = math.ceil((lastUpdate + UPDATE_TIMER) - time())
		UpdateTimer.text = string.format("UPDATES IN %s SECONDS", timeLeft)
	else
		lastUpdate = time()

		Update()
	end
end

------------------------------------------------------------------------------------------------------------------------
--	INITIALIZATION
------------------------------------------------------------------------------------------------------------------------

Events.Connect("LDT_Update", Update)
Events.Connect("TDT_Toggle", Toggle)

if(UPDATE_ON_ROUND_END) then
	Game.roundEndEvent:Connect(Update)
end

if(#UPDATE_ON_EVENT > 0) then
	Events.Connect(UPDATE_ON_EVENT, Update)
end

if(#TOGGLE_EVENT > 0) then
	Events.Connect(TOGGLE_EVENT, ForceToggle)
end

if(#FORCE_ON_EVENT > 0) then
	Events.Connect(FORCE_ON_EVENT, ForceOn)
end

if(#FORCE_OFF_EVENT > 0) then
	Events.Connect(TOGGLE_EVENT, ForceOff)
end

if(TOGGLE_BINDING) then
	LocalPlayer.bindingReleasedEvent:Connect(OnBindingReleased)
end

LEADERBOARD_TYPE = GetProperty(LEADERBOARD_TYPE, LEADERBOARD_TYPES)
LEADERBOARD_STAT = GetProperty(LEADERBOARD_STAT, LEADERBOARD_STATS)
LEADERBOARD_PERSISTENCE = GetProperty(LEADERBOARD_PERSISTENCE, LEADERBOARD_PERSISTENCES)
EASE_BEGINNING = GetProperty(EASE_BEGINNING, EASE_BEGINNINGS)

EASING_EQUATION_IN = EaseUI.EasingEquation[EASING_EQUATION_IN]
EASING_DIRECTION_IN = EaseUI.EasingEquation[EASING_DIRECTION_IN]
EASING_EQUATION_OUT = EaseUI.EasingEquation[EASING_EQUATION_OUT]
EASING_DIRECTION_OUT = EaseUI.EasingEquation[EASING_DIRECTION_OUT]

if(UPDATE_TIMER <= 0) then
	UPDATE_TIMER = 0
end

if(LEADERBOARD_STAT == "RESOURCE") then
	Title.text = string.format("%s %s %s", LEADERBOARD_TYPE, LEADERBOARD_PERSISTENCE, string.upper(RESOURCE_NAME))
else
	Title.text = string.format("%s %s %s", LEADERBOARD_TYPE, LEADERBOARD_PERSISTENCE, LEADERBOARD_STAT)
end

if(EASE_TOGGLE) then
	if(EASE_BEGINNING == "UP") then
		LeaderboardPanel.y = -1500
	elseif(EASE_BEGINNING == "DOWN") then
		LeaderboardPanel.y = 1500
	elseif(EASE_BEGINNING == "LEFT") then
		LeaderboardPanel.x = -1500
	elseif(EASE_BEGINNING == "RIGHT") then
		LeaderboardPanel.x = 1500
	end
end

while(not Leaderboards.HasLeaderboards()) do
	Task.Wait()
end
Task.Wait()
Update()